src/EVM/SMT.hs

```diff
@@ -811,6 +787,96 @@ getVars parseFn inst names = Map.mapKeys parseFn <$> foldM getOne mempty names
           r -> parseErr r
       pure $ Map.insert name val acc
 
+getStore :: SolverInstance -> [(Expr EWord, Expr EWord)] -> IO (Map W256 (Map W256 W256))
+getStore inst sreads = do
+  raw <- getValue inst "abstractStore"
+  let parsed = case parseCommentFreeFileMsg getValueRes (T.toStrict raw) of
+                 Right (ResSpecific (valParsed :| [])) -> valParsed
+                 r -> parseErr r
+      -- first interpret SMT term as a function
+      fun = case parsed of
+              (TermQualIdentifier (Unqualified (IdSymbol symbol)), term) ->
+                if symbol == "abstractStore"
+                then interpret2DArray Map.empty term
+                else error "Internal Error: solver did not return model for requested value"
+              r -> parseErr r
+
+  -- then create a map by adding only the locations that are read by the program
+  foldM (\m (addr, slot) -> do
+            addr' <- queryValue addr
+            slot' <- queryValue slot
+            pure $ addElem addr' slot' m fun) Map.empty sreads
+
+  where
+
+    addElem :: W256 -> W256 -> Map W256 (Map W256 W256) -> (W256 -> W256 -> W256) -> Map W256 (Map W256 W256)
+    addElem addr slot store fun =
+      case Map.lookup addr store of
+        Just m -> Map.insert addr (Map.insert slot (fun addr slot) m) store
+        Nothing -> Map.insert addr (Map.singleton slot (fun addr slot)) store
+
+
+    queryValue :: Expr EWord -> IO W256
+    queryValue (Lit w) = pure w
+    queryValue w = do
+      let expr = toLazyText $ exprToSMT w
+      raw <- getValue inst expr
+      case parseCommentFreeFileMsg getValueRes (T.toStrict raw) of
+        Right (ResSpecific (valParsed :| [])) ->
+          case valParsed of
+            (_, TermSpecConstant sc) -> pure $ parseW256 sc
+            _ -> error "Internal Error: cannot parse model for storage index"
+        r -> parseErr r
+
+
+
+-- | Interpret an N-dimentional array as a value of type a.
+-- Parameterized by an interpretation function for array values.
+interpretNDArray :: (Map Symbol Term -> Term -> a) -> (Map Symbol Term) -> Term -> (W256 -> a)
+interpretNDArray interp env = \case
+  -- variable reference
+  TermQualIdentifier (Unqualified (IdSymbol s)) ->
+    case Map.lookup s env of
+      Just t -> interpretNDArray interp env t
+      Nothing -> error "Internal error: unknown identifier, cannot parse array"
+  -- (let (x t') t)
+  TermLet (VarBinding x t' :| []) t -> interpretNDArray interp (Map.insert x t' env) t
+  TermLet (VarBinding x t' :| lets) t -> interpretNDArray interp (Map.insert x t' env) (TermLet (NonEmpty.fromList lets) t)
+  -- (as const (Array (_ BitVec 256) (_ BitVec 256))) SpecConstant
```

Comment:
Why is this commented out?

Reply:
these are comments denoting what is the SMT expression that this term corresponds too. Any idea how to make it clearer? For instance this node could correspond to a `((as const (Array (_ BitVec 256) (_ BitVec 256)))
  #x0000000000000000000000000000000000000000000000000000000000000001)`

Reply:
Ah, I see!!! No, that make sense, I didn't understand, I thought that was Haskell :D Sorry!